Homework1

ekhusainov opened the pull request:
Основные детали по проекту я написал в README.md.

Если кратко.

1. Данные игрушечные, поэтому для бейзлайна вполне хватило отчета pandas_profiling.
2. Написал логи для основного пайплайна. Их можно вывести в stderr, поменяв на yes тут в yml:
propagate: no
3. Категориальные признаки в one hot encoding. Числовые — нормировал.
4. Есть конфигурационные файлы для случайного леса и логистической регрессии (там ещё другой test_size и удалены некоторые признаки). 
```
python src\core.py fit_predict -c random_forest (по умолчанию)
python src\core.py fit_predict -c logregr
```
5. Основные python файлы:
```
core.py
build_features.py
fit_model.py
predict.py
all_train_params.py
```
6. Написал консольное приложение, поэтому predict реализовывается через тот же файл core.py.
```
python src\core.py predict

```
Выполнены почти все задания, подробнее в README.md

ekhusainov committed:
readme update

ekhusainov committed:
add readme

ekhusainov committed:
Create .gitignore

ekhusainov committed:
generate_structure

ekhusainov committed:
create report

ekhusainov committed:
logging

закончил писать логгирование для отчёта

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
Create setup.py

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
Update readme.md

ekhusainov committed:
Update readme.md

ekhusainov committed:
Update readme.md

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
fix logger build_features

ekhusainov committed:
Update build_features_logging.conf.yml

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
Update test_build_features.py

ekhusainov committed:
upd core work

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
Update fit_model.py

ekhusainov committed:
upd

ekhusainov committed:
there are accuracy score!

ekhusainov committed:
add features params

ekhusainov committed:
Update build_features.py

ekhusainov committed:
upd

ekhusainov committed:
Update train_fit.yml

ekhusainov committed:
config change works

ekhusainov committed:
upd

ekhusainov committed:
upd paths

ekhusainov committed:
upd paths

ekhusainov committed:
upd

ekhusainov committed:
pylint 10

ekhusainov committed:
predict plus

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
Update README.md

ekhusainov committed:
plus fake data

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
Create test_linter_ci.yml

ekhusainov committed:
Update test_linter_ci.yml

ekhusainov committed:
Update test_linter_ci.yml

ekhusainov committed:
upd

ekhusainov committed:
Create test_linter_ci.yml

ekhusainov committed:
Update test_linter_ci.yml

ekhusainov committed:
Update requirements.txt

ekhusainov committed:
upd

ekhusainov committed:
Update test_linter_ci.yml

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
Update test_core.py

ekhusainov committed:
Update test_linter_ci.yml

ekhusainov committed:
Update test_linter_ci.yml

ekhusainov committed:
upd

ekhusainov committed:
Update test_linter_ci.yml

ekhusainov committed:
upd

ekhusainov committed:
upd

ekhusainov committed:
Update README.md

ekhusainov committed:
complete01

ekhusainov committed:
Update README.md

ekhusainov committed:
Update README.md

ekhusainov committed:
Update README.md

ekhusainov committed:
Update README.md

ningeen commented on ml_project/README.md:
```diff
@@ -0,0 +1,181 @@
+│   └── fake_data      <- Сгенерированные данные.
+│
+│
+├── logs
```
без папки logs запуск core.py выдает ошибку

ekhusainov replied:
Спасибо, добавил
.gitkeep

ningeen commented on ml_project/src/core.py:
```diff
@@ -0,0 +1,114 @@
+    print(f"Accuracy score: {ac_score}")
+
+
+def callback_predict(argumets):
```
во время predict не вызывается функция setup_logging(), из-за чего логи не записываются

ekhusainov replied:
Переписал, добавил логгер в core.

ningeen commented on ml_project/report/creating_report.py:
```diff
@@ -0,0 +1,51 @@
+"""Here we just generate a report in html"""
```
наверное, для единства структуры проекта, файл с кодом лучше поместить в src/visualization

ekhusainov replied:
Спасибо, переместил

ningeen commented on ml_project/src/features/build_features.py:
```diff
@@ -0,0 +1,157 @@
+
+def split_to_train_test(data: pd.DataFrame,
+                        parametrs: TrainingPipelineParams,
+                        ) -> Tuple[pd.DataFrame, pd.DataFrame]:
```
Вроде бы тут надо так:
`) -> Tuple[pd.DataFrame, pd.DataFrame, pd.Series, pd.Series]:`

ekhusainov replied:
Done

ningeen commented on ml_project/src/features/build_features.py:
```diff
@@ -0,0 +1,157 @@
+
+def split_dataset_to_cat_num_features(x_data: pd.DataFrame,
+                                      parametrs: TrainingPipelineParams,
+                                      ) -> Tuple[pd.DataFrame]:
```
`) -> Tuple[pd.DataFrame, pd.DataFrame]:`

ekhusainov replied:
Done

ningeen commented on ml_project/src/features/build_features.py:
```diff
@@ -0,0 +1,157 @@
+
+
+if __name__ == "__main__":
+    main()
```
Не особо понял, зачем это надо, в других файлах тоже встречается. Мы же эту функцию отдельно не вызываем.

ekhusainov replied:
Спасибо, убрал.

ningeen commented on ml_project/src/core.py:
```diff
@@ -0,0 +1,114 @@
+
+def callback_predict(argumets):
+    "Argparse predict only."
+    parametrs = read_training_pipeline_params(DEFAULT_CONFIG_PATH)
```
predict настроен не очень гибко, здесь передаются параметры для тренировки случайного леса. если я обучаю логрег и пытаюсь запустить предикт - получаю ошибку.
Да и в целом, как по мне, предикт должен быть самостоятельной сущностью и от конфига трейна не зависеть.

ekhusainov replied:
Исправил ошибку следующим образом.
Сохраняю в папку models конфиг в joblib текущий, с помощью которого построена модель, что важно, так как в конфиге может меняться количество фичей. Пока не очень очевидно, как сделать так, чтобы от конфига трейна не зависело. В конфиг трейне может быть 2 фичи только, а тест будет подаваться с полными, причём модель не будет знать какие именно фичи использовались по умолчанию. Здесь сохранённый конгфиг использую, чтобы вытащить нужные 2 фичи и делать предикт только по ним.

ningeen replied:
Можно во время трейна создавать нужный пайплайн с препроцессингом и моделью и его сохранять. А во время инференса просто его загружать и подавать ему на вход сырые данные, он уже будет знать как ему их обработать. Поэтому я ниже и упоминал pipeline & transformers

ningeen commented on ml_project/src/features/build_features.py:
```diff
@@ -0,0 +1,157 @@
+    logger.info("Finish saving transformer to %s.", repr(filepath))
+
+
+def build_features(parametrs: TrainingPipelineParams,
```
В коде от препода был отличный пример использования pipeline & transformers. Очень удобный инструмент, зачем было переписывать - непонятно.
https://scikit-learn.org/stable/modules/compose.html#combining-estimators

ekhusainov replied:
Спасибо, действительно очень удобный инструмент и сильно сократит мой код. А то я старался следовать указанием к описанию ДЗ, поэтому всё писал сам)
"Пример подобного разбирали на паре https://github.com/made-ml-in-prod-2021/ml_project_example, **не нужно отсюда копировать**=)"
Дождусь ревью преподавателя, готов стерпеть небольшие штрафы по баллам за мой код(

ol-ga-00 replied:
@ekhusainov, считаю, что ваш способ реализации build_features менее гибкий, чем в примере, разбираемом на паре в https://github.com/made-ml-in-prod-2021/ml_project_example (изменение препроцессинга потребует серьёзные доработки). 
Также рекомендую использовать Pipeline, что указал ningeen!

Но ваш код имеет право на жизнь, снижать баллы за это не буду

ekhusainov replied:
Спасибо, уже хорошо это осознал, пока этот код в докер оборачивал))

ningeen commented on ml_project/src/features/build_features.py:
```diff
@@ -0,0 +1,157 @@
+
+APPLICATION_NAME = "build_features"
+DEFAULT_LOGGING_PATH = "configs/core_logging.conf.yml"
+PATH_TO_ONE_HOT_ENCODER = "models/one_hot.joblib"
+PATH_TO_SCALER = "models/standart_scaler.joblib"
+PREPROCESSED_DATA_FILEPATH = "data/processed/x_train_for_fit_predict.csv"
+X_TEST_FILEPATH = "data/validate_part/x_test.csv"
+Y_TEST_FILEPATH = "data/validate_part/y_test.csv"
+Y_TRAIN_FILEPATH = "data/processed/y_train.csv"
```
Почему бы такие вещи не вынести в конфиг?

ekhusainov replied:
было больно, так как этого много, но сделал)

ningeen commented on ml_project/report/creating_report.py:
```diff
@@ -0,0 +1,51 @@
+
+
+def setup_logging():
+    "Logger from yaml config."
```
Для docstring обычно исползуются тройные кавычки

ekhusainov replied:
Сорян, мимо меня прошёл этот пункт PEP 257, поправил.

ningeen reviewed (changes requested):
В целом, если не душнить, то оцениваю в 32 балла.
Если душнить, то отнял бы балл за отсутствие EDA в виде ноутбука и еще балл за не совсем корректную реализацию predict, итого 30.

ekhusainov committed:
upd

ekhusainov committed:
upd_structure

report to visualization

ekhusainov committed:
upd predict

ekhusainov committed:
fix

ekhusainov committed:
plus quotes

ekhusainov committed:
tests work

ekhusainov committed:
Update core.py

ekhusainov committed:
pylint 10

ekhusainov committed:
upd report

ekhusainov committed:
finish

ekhusainov committed:
Update README.md

ekhusainov committed:
add score in formatter logs

Теперь с квадратными скобками
[2021-05-10 16:27:26] build_features [INFO] Start reading the file.

ekhusainov committed:
Revert "add score in formatter logs"

This reverts commit 52ed633e65be53ae4799a7e81e010dc761b6fc19.

ekhusainov committed:
upd

ol-ga-00 commented on ml_project/src/core.py:
```diff
@@ -0,0 +1,137 @@
+DEFAULT_CONFIG_NAME = "random_forest"
+DEFAULT_CONFIG_PATH = "configs/random_forest.yml"
+DEFAULT_DATASET_FOR_PREDICT = "data/validate_part/x_test.csv"
+DEFAULT_PREDICTED_DATA = "data/y_pred/y_pred.csv"
```
В названиях есть слово "DEFAULT_", что намекает на то, что эти переменные можно будет переопределить. Но по коду я не поняла, как это можно сделать.
Думаю, что их сейчас реально нельзя переопределить.
Чтобы это можно было сделать, рекомендую вынести их в конфиг!

ekhusainov replied:
Первые два и связаны с конфигом, поэтому отдельно не засунешь в конфиг. Убрал DEFAULT, спасибо, не знал про это.

ol-ga-00 commented on ml_project/tests/test_fit_model.py:
```diff
@@ -0,0 +1,15 @@
+    # DEFAULT_Y_TRAIN_PATH,
+    read_csv_file,
+)
+from src.core import DEFAULT_CONFIG_PATH
```
src/ и tests/ должны быть независимы. 
Из src/ необходимо импортировать только то, что будет тестироваться. 
Константы же импортировать оттуда плохо (да и вообще их все нужно вынести в конфиг)

Рекомендую создать конфиг для тестов. Положить его можно прям в tests/

ekhusainov replied:
Извиняюсь, не сразу заметил этот коммент (плюс выше кликну на Re-request review, только сейчас заметил).
Поправил, удалил импорт константы

ekhusainov commented:
@ol-ga-00 Можно уже делать merge?

ol-ga-00 reviewed (commented):
> @ol-ga-00 Можно уже делать merge?

нужно дождаться моего апрува

ol-ga-00 reviewed (changes requested):
-2) Назовите ветку homework1 (1 балл)  -- 1

➕ -1) положите код в папку ml_project -- 0

➕ 0) В описании к пулл реквесту описаны основные "архитектурные" и тактические решения, которые сделаны в вашей работе. В общем, описание что именно вы сделали и для чего, чтобы вашим ревьюерам было легче понять ваш код. (2 балла)  -- 2

➕ 1) Выполнение EDA, закоммитьте ноутбук в папку с ноутбуками (2 баллов) -- 0
> не нашла ноутбук
Вы так же можете построить в ноутбуке прототип(если это вписывается в ваш стиль работы) Можете использовать не ноутбук, а скрипт, который сгенерит отчет, закоммитьте и скрипт и отчет (за это + 1 балл) -- 1

➕ 2) Проект имеет модульную структуру(не все в одном файле =) ) (2 баллов) -- 1 
> Код разнесён на смысловые директории (это плюс). Но каждый этап (препроцессинг, обучение, предсказание) зависит от предыдущих, что нарушает модульность. В тестах это чётко видно. Чтобы протестировать какую-нибудь функцию, приходится запускать код всего, что в реальности будет выполняться до её вызова (это минус). 

➕ 3) использованы логгеры (2 балла) -- 2

➕ 4) написаны тесты на отдельные модули и на прогон всего пайплайна(3 баллов) -- 2
> Плохо завязывать тесты на константы основного кода, всё необходимо вынести в конфиг (я про DEFAULT_CONFIG_PATH). 

➕ 5) Для тестов генерируются синтетические данные, приближенные к реальным (3 баллов) -- 0
можно посмотреть на библиотеки https://faker.readthedocs.io/en/, https://feature-forge.readthedocs.io/en/latest/
можно просто руками посоздавать данных, собственноручно написанными функциями как альтернатива, можно закоммитить файл с подмножеством трейна(это не оценивается)
> синтетические данные создаются в test_core.py, но нигде не используются!

➕ 6) Обучение модели конфигурируется с помощью конфигов в json или yaml, закоммитьте как минимум 2 корректные конфигурации, с помощью которых можно обучить модель (разные модели, стратегии split, preprocessing) (3 балла) -- 3

➕ 7) Используются датаклассы для сущностей из конфига, а не голые dict (3 балла) -- 3

➖ 8) Используйте кастомный трансформер(написанный своими руками) и протестируйте его(3 балла) -- 0

➕ 9) Обучите модель, запишите в readme как это предлагается (3 балла) -- 3

➕ 10) напишите функцию predict, которая примет на вход артефакт/ы от обучения, тестовую выборку(без меток) и запишет предикт, напишите в readme как это сделать (3 балла) -- 3

➖ 11) Используется hydra (https://hydra.cc/docs/intro/) (3 балла - доп баллы) -- 0

➕ 12) Настроен CI(прогон тестов, линтера) на основе github actions (3 балла - доп баллы (будем проходить дальше в курсе, но если есть желание поразбираться - welcome) -- 3

➕ 13) Проведите самооценку, опишите, в какое колво баллов по вашему мнению стоит оценить вашу работу и почему (1 балл доп баллы)  -- 1


Насчитала 25 баллов
@ekhusainov, согласны с моей оценкой или хотите обсудить некоторые пункты?

ekhusainov commented:
> ➕ 1) Выполнение EDA, закоммитьте ноутбук в папку с ноутбуками (2 баллов) -- 0
> 
> > не нашла ноутбук
> > Вы так же можете построить в ноутбуке прототип(если это вписывается в ваш стиль работы) Можете использовать не ноутбук, а скрипт, который сгенерит отчет, закоммитьте и скрипт и отчет (за это + 1 балл) -- 1

С этим пунктом не согласен. Я использовал скрипт, который сгенерировал отчёт, на основе pandas-profiling. Формулировку "(за это +1 балл)" выглядит так, что за такое 2+1 балла. Если бы было "(за это 1 балл)", то выглядело бы как 1 балл. Возможно у нас какие-то диалектические расхождения, но в данный момент почему-то стоит 0 баллов. Тем более данные такие игрушечные, что я посчитал, что отчёта данного должно хватить для уверенного бейзлайна...

> ➕ 5) Для тестов генерируются синтетические данные, приближенные к реальным (3 баллов) -- 0
> можно посмотреть на библиотеки https://faker.readthedocs.io/en/, https://feature-forge.readthedocs.io/en/latest/
> можно просто руками посоздавать данных, собственноручно написанными функциями как альтернатива, можно закоммитить файл с подмножеством трейна(это не оценивается)
> 
> > синтетические данные создаются в test_core.py, но нигде не используются!
> 
Возможно не понял это задание, можете пояснить? Я же генерировал для тестов данные, их надо было просто сохранять в tmp_path? за это минус 3 балла?
UPD: Сейчас пришло в голову. Надо было сгенерированные данные потом во всех тестах использовать?


С остальным согласен.
P.S. Ну, и там к этим баллам надо добавить баллы, за проделанное ревью.
https://github.com/made-ml-in-prod-2021/asv715/pull/3

ekhusainov commented:
Походу забыл пинг кинуть, к моему комменту последнему.
@ol-ga-00

ekhusainov committed:
upd

ekhusainov committed:
- DEFAULT

ol-ga-00 commented:
> > ➕ 1) Выполнение EDA, закоммитьте ноутбук в папку с ноутбуками (2 баллов) -- 0
> > > не нашла ноутбук
> > > Вы так же можете построить в ноутбуке прототип(если это вписывается в ваш стиль работы) Можете использовать не ноутбук, а скрипт, который сгенерит отчет, закоммитьте и скрипт и отчет (за это + 1 балл) -- 1
> 
> С этим пунктом не согласен. Я использовал скрипт, который сгенерировал отчёт, на основе pandas-profiling. Формулировку "(за это +1 балл)" выглядит так, что за такое 2+1 балла. Если бы было "(за это 1 балл)", то выглядело бы как 1 балл. Возможно у нас какие-то диалектические расхождения, но в данный момент почему-то стоит 0 баллов. Тем более данные такие игрушечные, что я посчитал, что отчёта данного должно хватить для уверенного бейзлайна...
> 
Вы правы! За пункт 1) поставлю 3 балла
> > ➕ 5) Для тестов генерируются синтетические данные, приближенные к реальным (3 баллов) -- 0
> > можно посмотреть на библиотеки https://faker.readthedocs.io/en/, https://feature-forge.readthedocs.io/en/latest/
> > можно просто руками посоздавать данных, собственноручно написанными функциями как альтернатива, можно закоммитить файл с подмножеством трейна(это не оценивается)
> > > синтетические данные создаются в test_core.py, но нигде не используются!
> 
> Возможно не понял это задание, можете пояснить? Я же генерировал для тестов данные, их надо было просто сохранять в tmp_path? за это минус 3 балла?
> UPD: Сейчас пришло в голову. Надо было сгенерированные данные потом во всех тестах использовать?

Да, сгенерированные данные нужно было использовать во всех тестах

> 
> С остальным согласен.
> P.S. Ну, и там к этим баллам надо добавить баллы, за проделанное ревью.
> [made-ml-in-prod-2021/asv715#3](https://github.com/made-ml-in-prod-2021/asv715/pull/3)

Это отдельно будет учитываться

ol-ga-00 reviewed (approved):
Ставлю 27 баллов!

ekhusainov commented:
> 
> 
> Ставлю 27 баллов!

Можно делать Merge?)

ol-ga-00 commented:
> > Ставлю 27 баллов!
> 
> Можно делать Merge?)

Можно!
